base/dataset_prepare.go

```diff
@@ -12,66 +13,117 @@ import (
 	"github.com/qri-io/dataset/detect"
 	"github.com/qri-io/dataset/validate"
 	"github.com/qri-io/qfs"
-	"github.com/qri-io/qri/base/dsfs"
 	"github.com/qri-io/qri/dsref"
-	"github.com/qri-io/qri/repo"
+	qerr "github.com/qri-io/qri/errors"
+	"github.com/qri-io/qri/logbook"
 	"github.com/qri-io/qri/repo/profile"
-	reporef "github.com/qri-io/qri/repo/ref"
 )
 
-// PrepareHeadDatasetVersion prepares to save by loading the head commit, opening the body file,
-// and constructing a mutable version that has no transform or commit.
-func PrepareHeadDatasetVersion(ctx context.Context, r repo.Repo, peername, name string) (curr, mutable *dataset.Dataset, currPath string, err error) {
-	// TODO(dustmop): We should not be calling CanonicalizeDatasetRef here. It's already been
-	// called up in lib, which means that we've thrown information away. Furthermore, we
-	// should be relying on stable identifiers this low down the stack. Instead, load the dataset
-	// head (if it exists) by using the initID, and pass it into base.Save.
-
-	// Determine if the save is creating a new dataset or updating an existing dataset by
-	// seeing if the name can canonicalize to a repo that we know about
-	lookup := &reporef.DatasetRef{Name: name, Peername: peername}
-	if err = repo.CanonicalizeDatasetRef(r, lookup); err == repo.ErrNotFound || lookup.Path == "" {
-		return &dataset.Dataset{}, &dataset.Dataset{}, "", nil
+// PrepareSaveRef works out a dataset reference for saving a dataset
+// version. When a dataset exists, resolve the initID & path
+// When no history exists, ensure a locally-unique dataset name and create a
+// new logbook history & InitID to write to
+// successful calls to PrepareSaving reference always have an InitID, and may
+// have a Path value
+func PrepareSaveRef(
+	ctx context.Context,
+	pro *profile.Profile,
+	book *logbook.Book,
+	resolver dsref.Resolver,
+	refStr string,
+	bodyPathStr string,
+	newName bool,
```

Comment:
I don't think `newName` differentiates this value well (especially since the introduction of `isNew`), it's unclear exactly what it means. That is, it represents if the user is explicitly requesting a new dataset be created, whose name is not in use yet. I would rename it to `wantNewName` or `requestingNewName`.